Lib/posixpath.py

```diff
@@ -436,8 +439,10 @@ def realpath(filename, *, strict=False):
         else:
             newpath = path + sep + name
         try:
-            st = os.lstat(newpath)
-            if not stat.S_ISLNK(st.st_mode):
+            st_mode = os.lstat(newpath).st_mode
+            if not stat.S_ISLNK(st_mode):
+                if strict and part_count and not stat.S_ISDIR(st_mode):
+                    raise NotADirectoryError(errno.ENOTDIR, "Not a directory", newpath)
```

Comment:
There's no need to explicitly raise `NotADirectoryError` if `errno` is `ENOTDIR`. Instantiating `OSError` substitutes the relevant subclass automatically:

```pycon
>>> OSError(errno.ENOTDIR, '')
NotADirectoryError(20, '')
```

Also, the error message shouldn't be hard coded, and PEP 8 limits line length to 79 characters.

```suggestion
                    raise OSError(errno.ENOTDIR, os.strerror(errno.ENOTDIR),
                                  newpath)
```